### PR TITLE
Added categories

### PR DESCRIPTION
**Pull request by VigneshPalanisamy**

### CODE REVIEW

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+// author: Vignesh Palanisamy
+// version: 2.1.0
+//
+
+import UIKit
+
+class CategoriesTableViewController: UITableViewController   {
+    
+    var categoryModel = [Category]()
+    var categoryName: String = ""
```
Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: Why don't define as private?

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    fileprivate var songTabBarController: SongsTabBarViewController?
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        songTabBarController = self.tabBarController as? SongsTabBarViewController
+        self.tabBarItem.title = "categories".localized
+        tableView.contentInset = UIEdgeInsetsMake(0, 0, (self.tabBarController?.tabBar.frame.height)!, 0)
+        //refresh control
+        refresh = UIRefreshControl()
+        refresh.attributedTitle = NSAttributedString(string: "Refresh")
```
Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: Why don't use localized here?

Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: can you move ui refresh relevant code separate method?

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    
+    var searchBar: UISearchBar!
+    var refresh = UIRefreshControl()
+    fileprivate var songTabBarController: SongsTabBarViewController?
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        songTabBarController = self.tabBarController as? SongsTabBarViewController
+        self.tabBarItem.title = "categories".localized
+        tableView.contentInset = UIEdgeInsetsMake(0, 0, (self.tabBarController?.tabBar.frame.height)!, 0)
```
Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: is this really needed?

Comment (VigneshPalanisamy) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: yes it is required..

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    var songName: String = ""
+    
+    var searchBar: UISearchBar!
+    var refresh = UIRefreshControl()
+    fileprivate var songTabBarController: SongsTabBarViewController?
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        songTabBarController = self.tabBarController as? SongsTabBarViewController
+        self.tabBarItem.title = "categories".localized
```
Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: Repeated in viewWillAppear method.
can you keep in common method?

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        songTabBarController.navigationItem.title = "categories".localized
+        categoryModel = databaseHelper.getCategoryModel()
+        filteredCategoryModel = categoryModel
+        createSearchBar()
+        tableView.reloadData()
+    }
+    
+    override func didReceiveMemoryWarning() {
+        super.didReceiveMemoryWarning()
+        // Dispose of any resources that can be recreated.
```
Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: Remove this commented code

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    }
+    
+    // MARK: - Table view data source
+    
+    override func numberOfSections(in tableView: UITableView) -> Int {
+        // #warning Incomplete implementation, return the number of sections
```
Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: Remove this commented code

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        return 1
+    }
+    
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        // #warning Incomplete implementation, return the number of rows
+        return filteredCategoryModel.count
```
Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: Remove this commented code

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    
+    
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
+        cell.textLabel?.text = filteredCategoryModel[(indexPath as NSIndexPath).row].name
+        return cell
+    }
+    
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+
```
Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: Remove empty lines

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any!) {
+        if (segue.identifier == "artistTitle") {
+            let titleTableViewController = segue.destination as! ArtistSongsTitleTableViewController
+            titleTableViewController.artistName = categoryName
+            titleTableViewController.songModel = songsModel
+            titleTableViewController.songTabBarController = songTabBarController
+        }
+    }
+
```
Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: Remove empty lines

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift (hunk 2 above)
Comment (VigneshPalanisamy) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: why we need separate method.. it was not called anywhere else.

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    }
+    
+}
+
+extension CategoriesTableViewController: UISearchBarDelegate {
+    
+    func createSearchBar()
+    {
+        // Search bar
+        let searchBarFrame = CGRect(x: self.view.bounds.origin.x, y: self.view.bounds.origin.y, width: self.view.bounds.size.width, height: 44);
```
Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: Is this comment needed?

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        self.tabBarController?.navigationItem.setRightBarButton(UIBarButtonItem(barButtonSystemItem: .search, target: self, action: #selector(ArtistsTableViewController.searchButtonItemClicked(_:))), animated: true)
+    }
+    
+    func searchButtonItemClicked(_ sender:UIBarButtonItem){
+        self.tabBarController?.navigationItem.titleView = searchBar;
+        self.tabBarController?.navigationItem.leftBarButtonItem?.isEnabled = false
+        self.tabBarController?.navigationItem.rightBarButtonItem = nil
+        searchBar.becomeFirstResponder()
+    }
+
```
Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: Remove empty lines

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+        self.tabBarController?.navigationItem.setRightBarButton(UIBarButtonItem(barButtonSystemItem: .search, target: self, action: #selector(ArtistsTableViewController.searchButtonItemClicked(_:))), animated: true)
+    }
+    
+    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
+        filterContentForSearchText(self.searchBar)
+        self.tableView.reloadData()
+    }
+    
+    func filterContentForSearchText(_ searchBar: UISearchBar) {
+        // Filter the array using the filter method
```
Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: Some method Braces started next line, here end of method.
Use anyone convention

File: worshipsongs/worshipsongs/Category.swift [hunk 13]
```diff
@@ -0,0 +1,14 @@
+//
+//  Category.swift
+//  worshipsongs
+//
+//  Created by Vignesh Palanisamy on 31/03/2017.
+//  Copyright © 2017 Vignesh Palanisamy. All rights reserved.
+//
+
```
Comment (madasamys) [inline worshipsongs/worshipsongs/Category.swift]: Add proper author and version

File: worshipsongs/worshipsongs/Category.swift [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+import Foundation
+
+struct Category {
+    var id: String
```
Comment (madasamys) [inline worshipsongs/worshipsongs/Category.swift]: Usually id in long or int in db. 
Is there any reason to define as String?

File: worshipsongs/worshipsongs/DatabaseHelper.swift [hunk 15]
```diff
@@ -192,4 +192,61 @@ class DatabaseHelper: NSObject {
         return authorName
     }
     
+    func getCategoryModel() -> [(Category)] {
+        database = FMDatabase(path: commonService.getDocumentDirectoryPath("songs.sqlite"))
+        let path = commonService.getDocumentDirectoryPath("songs.sqlite")
+        //54D70B97-F386-4746-9A69-692E339668B8
```
Comment (madasamys) [inline worshipsongs/worshipsongs/DatabaseHelper.swift]: Do we need this?

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift (hunk 4 above)
Comment (VigneshPalanisamy) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: can you please explain what are the repeated things

File: worshipsongs/worshipsongs/DatabaseHelper.swift [hunk 16]
```diff
@@ -192,4 +192,61 @@ class DatabaseHelper: NSObject {
...
+                if(args != "")
+                {
+                    args="\(args),"
+                }
+                args="\(args)?"
+                songModelIds.append(resultSet1!.string(forColumn: "song_id") as AnyObject)
+            }
+        }
+        var songModel = [Songs]()
+        let resultSet2: FMResultSet? = database!.executeQuery("SELECT * FROM songs where id IN (\(args)) ORDER BY title", withArgumentsIn: songModelIds)
```
Comment (madasamys) [inline worshipsongs/worshipsongs/DatabaseHelper.swift]: This is not effective to constructing two query to find song model for respective category. 

Why don't use inner join like below?
`select * from songs as s inner join songs_topics as st on st.song_id = s.id inner join topics as t on st.topic_id = t.id where t.id= ? ORDER by title`

File: worshipsongs/worshipsongs/DatabaseHelper.swift [hunk 17]
```diff
@@ -192,4 +192,61 @@ class DatabaseHelper: NSObject {
...
+        {
+            while resultSet1!.next() {
+                categoryModel.append(Category(id: resultSet1!.string(forColumn: id), name: resultSet1!.string(forColumn: name)))
+            }
+        }
+        print("songModel count : \(categoryModel.count)")
+        return categoryModel
+    }
+    
+    func getCategorySongsModel(_ argument: String) -> [(Songs)] {
```
Comment (madasamys) [inline worshipsongs/worshipsongs/DatabaseHelper.swift]: Rename to findCategorySongs(...)

File: worshipsongs/worshipsongs/DatabaseHelper.swift [hunk 18]
```diff
@@ -192,4 +192,61 @@ class DatabaseHelper: NSObject {
         return authorName
     }
     
+    func getCategoryModel() -> [(Category)] {
```
Comment (madasamys) [inline worshipsongs/worshipsongs/DatabaseHelper.swift]: findCategories()

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift (hunk 4 above)
Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: self.tabBarItem.title = "categories".localized

File: worshipsongs/worshipsongs/CategoriesTableViewController.swift (hunk 2 above)
Comment (madasamys) [inline worshipsongs/worshipsongs/CategoriesTableViewController.swift]: Code should be readable, keep in separate method.

File: worshipsongs/worshipsongs/DatabaseHelper.swift (hunk 16 above)
Comment (VigneshPalanisamy) [inline worshipsongs/worshipsongs/DatabaseHelper.swift]: it's ok to have two queries... it was much more effective than the inner join.

Comment (madasamys) [inline worshipsongs/worshipsongs/DatabaseHelper.swift]: Executing two queries not much more effective than inner join.

**Pull request closed without merging**